addon/model.js

```diff
@@ -260,55 +251,146 @@ export default class MegamorphicModel extends Ember.Object {
     return this._internalModel.modelName;
   }
 
+  get _isProjection() {
+    return this._baseModel != null;
+  }
+
   __defineNonEnumerable(property) {
     this[property.name] = property.descriptor.value;
   }
 
-  _assignAttributes(attributes) {
-    // Don't merge; overwrite
-    this._internalModel._data = attributes;
+  _assignAttributes() {
+    // do not do anything as changed keys will apply the changes
+    // to discover the changed ones
```

Comment:
This is only done to avoid traversing the graph of changes twice and it is limitation of the existing API (`_changedKeys` and `_assignAttributes`). Apart from this, I totally agree with you.

Reply:
@dnachev 👍 

let's revisit this once we're using modeldata api and see if there's still an issue